Selenium doesn't need MCP

p0deje opened the pull request:
### **User description**
<!--- Provide a general summary of your changes in the Title above -->

### Description

Adds a blog post to clarify the Selenium project's stance on whether it should include an official MCP server.

### Motivation and Context

It's been repeatedly asked in bug trackers and Slack channels.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Change to the site (I have double-checked the Netlify deployment, and my changes look good)
- [ ] Code example added (and I also added the example to all translated languages)
- [ ] Improved translation
- [ ] Added new translation (and I also added a notice to each document missing translation)

### Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] I have read the [**contributing**](https://selenium.dev/documentation/en/contributing/) document.
- [x] I have used [hugo](https://gohugo.io/) to render the site/docs locally and I am sure it works.
<!--- Provide a general summary of your changes in the Title above -->


___

### **PR Type**
Documentation


___

### **Description**
- Adds blog post explaining Selenium's stance on MCP server implementation

- Discusses context window limitations and token consumption issues

- Compares tool calling vs code execution approaches for AI agents

- Advocates for ecosystem-based solutions over native MCP integration


___

### Diagram Walkthrough


```mermaid
flowchart LR
  A["Selenium Project"] -->|"Clarifies stance"| B["MCP Server Question"]
  B -->|"Identifies problems"| C["Context Window Limitations"]
  C -->|"Proposes alternative"| D["Ecosystem-based Solutions"]
  D -->|"Recommends"| E["Code Execution via Bindings"]
```



<details><summary><h3>File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>why-selenium-doesnt-need-mcp.md</strong><dd><code>Blog post on Selenium MCP server rationale</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

website_and_docs/content/blog/2025/why-selenium-doesnt-need-mcp.md

<ul><li>New blog post addressing recurring community question about official <br>MCP server<br> <li> Explains why native MCP integration is inefficient for LLM agents<br> <li> Details context window management challenges with naive tool exposure<br> <li> Compares tool calling vs code execution paradigms for agent automation<br> <li> Advocates for ecosystem projects to handle MCP implementation instead</ul>


</details>


  </td>
  <td><a href="https://github.com/SeleniumHQ/seleniumhq.github.io/pull/2558/files#diff-fc19fc3293b10fc4443866467b17718844f9c53e5503111d5faecc6d69b13fe8">+77/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tbody></table>

</details>

___

p0deje committed:
Why Selenium doesn't need MCP

titusfortner commented on website_and_docs/content/blog/2025/why-selenium-doesnt-need-mcp.md:
```diff
@@ -0,0 +1,77 @@
+---
+title: Why Selenium Doesn't Need MCP
```
```suggestion
title: Selenium Doesn't Need MCP
```

titusfortner commented on website_and_docs/content/blog/2025/why-selenium-doesnt-need-mcp.md:
```diff
@@ -0,0 +1,77 @@
+---
+title: Why Selenium Doesn't Need MCP
+linkTitle: Why Selenium Doesn't Need MCP
```
```suggestion
linkTitle: Selenium Doesn't Need MCP
```

titusfortner commented on website_and_docs/content/blog/2025/why-selenium-doesnt-need-mcp.md:
```diff
@@ -0,0 +1,77 @@
+categories: ["technical"]
+author: Alex Rodionov ([@p0deje](https://twitter.com/p0deje))
+description: >
+   Selenium project stance on whether it should have an official MCP server
```
```suggestion
   Technical explanation of why we won't create an official MCP Server
```

titusfortner commented on website_and_docs/content/blog/2025/why-selenium-doesnt-need-mcp.md:
```diff
@@ -0,0 +1,77 @@
+
+The most common request we see is for a simple MCP server that exposes our API - `find_element`, `click`, `get_page_source` - as tools. While this sounds useful, in practice, it leads to immediate failure modes regarding _context window management_. 
+
+We can look at early adopters for evidence. The official GitHub MCP server provides a wide array of tools, which results in massive context consumption. Its initial implementation exposed 100+ tools that consumed 64,000 tokens. That’s before the agent does any work. This issue was [improved](https://github.com/github/github-mcp-server/discussions/1182) over time, but even today, the server still  ...
```
I don't have a good reference for the token values you've included. Is there a recommended limit for what this should be? 
This might be assuming the users know more than they do. Maybe highlight that these tokens are added to the context window for every call. 

Actually, can make this use case more relevant by framing it that people looking to use LLM clients for browser testing will want to also load the Jira MCP for logging bugs, and the GitHub MCP for raising issues or fixing test code.

titusfortner replied:
Maybe pull the 200k tokens guideline link up to this paragraph, and reference the number again below.

p0deje replied:
I've added a line saying that Opus 4.5 context window is 200k

titusfortner reviewed (dismissed):
This is excellent.

I gave suggestions for making the title/subtitle stronger, and then more feedback on the GitHub paragraph for consideration.

p0deje committed:
Update website_and_docs/content/blog/2025/why-selenium-doesnt-need-mcp.md

Co-authored-by: Titus Fortner <titusfortner@users.noreply.github.com>

p0deje committed:
Update website_and_docs/content/blog/2025/why-selenium-doesnt-need-mcp.md

Co-authored-by: Titus Fortner <titusfortner@users.noreply.github.com>

p0deje committed:
Update website_and_docs/content/blog/2025/why-selenium-doesnt-need-mcp.md

Co-authored-by: Titus Fortner <titusfortner@users.noreply.github.com>

p0deje committed:
Revise blog post on Selenium and MCP context issues

Clarify the limitations of MCP in context management and provide evidence from early adopters.

p0deje committed:
add link

rpallavisharma commented:
Informative blog post. Thanks for writing and explaining Alex @p0deje . places where it says ours we may say Selenium or
Project. 

And I just have a question the article ends with saying once the problems with mcp are solved we might see what to do.. 

selenium project has ide , has grid these are beyond the WebDriver bidi low level implementations. 
Similarly selenium manager came into being.. 

Is it better to say why there is nothing official from project yet and have that approach because of the challenges and technicalities the blog mentions? That's just a thought that I had. 

Just a thought to keep it open ended maybe.. as project decides... how to proceed further ..

p0deje commented:
@rpallavisharma Thank you for taking the time to read it and share your feedback. I'm planning to significantly rewrite this based on this and other folks comments,
